### PR TITLE
Clang compiler

### PR DESCRIPTION
**Pull request by NathanYukai**
implemented clang compiler option, updated readme for clang-omp

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,8 +49,9 @@ environment. To execute the above test case in parallel run:
 ```
 python tests/eigenwave3d.py --compiler <cc> --execute --nthreads <nt>
 ```
-where `<cc>` is either `g++` or `icpc` and `<nt>` is the number of
-threads to use during execution. For additional options please see:
+where `<cc>` is either `g++`,`clang` or `icpc` and `<nt>` is the number of
+threads to use during execution. Make sure your [clang](http://clang-omp.github.io/) compiler support openmp
```
Comment (ctjacobs) [inline README.md]: "support" -> "supports"

File: README.md [hunk 2]
```diff
@@ -49,8 +49,9 @@ environment. To execute the above test case in parallel run:
 ```
 python tests/eigenwave3d.py --compiler <cc> --execute --nthreads <nt>
 ```
-where `<cc>` is either `g++` or `icpc` and `<nt>` is the number of
-threads to use during execution. For additional options please see:
+where `<cc>` is either `g++`,`clang` or `icpc` and `<nt>` is the number of
+threads to use during execution. Make sure your [clang](http://clang-omp.github.io/) compiler support openmp 
+for multithreads program. For additional options please see:
```
Comment (ctjacobs) [inline README.md]: "multithreads program" -> "multi-threaded programs"

File: opesci/compilation.py [hunk 3]
```diff
@@ -64,6 +64,16 @@ def __init__(self, cppargs=[], ldargs=[]):
     def _ivdep(self):
         return '#pragma GCC ivdep'
 
+class ClangCompiler(Compiler):
+    def __init__(self,cppargs=[],ldargs=[]):
```
Comment (ctjacobs) [inline opesci/compilation.py]: Add docstrings please - see the `GNUCompiler` class for inspiration.

**Comment by mlange05** [comment]
I think there are still some PEP8 failures in this branch that want fixing. I think the changes look fine otherwise, although I haven't tested the threaded runs with `clang-omp`. Could anyone with a Mac please try the new compiler option?

**Comment by ggorman** [comment]
Isn’t there a clang option on Travis?

On 8 Sep 2015, at 14:35, Michael Lange <notifications@github.com<mailto:notifications@github.com>> wrote:


I think there are still some PEP8 failures in this branch that want fixing. I think the changes look fine otherwise, although I haven't tested the threaded runs with clang-omp. Could anyone with a Mac please try the new compiler option?

**Comment by mlange05** [comment]
Yes, there is. We should run out test suite through both compiler setups. @NathanYukai, can you add this please?

**Comment by ggorman** [comment]
I'm still seeing a lot of problems:

src/eigenwave3d.cpp:168:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 2);
      ^
src/eigenwave3d.cpp:166:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 2);
      ^
src/eigenwave3d.cpp:164:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 2);
      ^
src/eigenwave3d.cpp:179:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 2);
      ^
src/eigenwave3d.cpp:177:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 2);
      ^
src/eigenwave3d.cpp:175:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 2);
      ^
src/eigenwave3d.cpp:190:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 2);
      ^
src/eigenwave3d.cpp:188:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 2);
      ^
src/eigenwave3d.cpp:186:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 2);
      ^
src/eigenwave3d.cpp:201:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 2);
      ^
src/eigenwave3d.cpp:199:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 1.5);
      ^
src/eigenwave3d.cpp:197:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 1.5);
      ^
src/eigenwave3d.cpp:212:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 1.5);
      ^
src/eigenwave3d.cpp:210:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 1.5);
      ^
src/eigenwave3d.cpp:208:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 2);
      ^
src/eigenwave3d.cpp:223:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 1.5);
      ^
src/eigenwave3d.cpp:221:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 2);
      ^
src/eigenwave3d.cpp:219:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 1.5);
      ^
src/eigenwave3d.cpp:264:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:272:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:280:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:288:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:296:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:304:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:312:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:320:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:328:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:336:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:344:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:352:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:360:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:368:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:376:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:384:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:392:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:400:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:408:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:415:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:422:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:429:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:436:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:443:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:450:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:457:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:464:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:471:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:478:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:485:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:492:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:499:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:506:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:513:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:520:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:527:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:546:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:560:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:568:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:576:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:584:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:592:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:600:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:608:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:616:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:624:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:632:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:640:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:648:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:656:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:664:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:672:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:680:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:688:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:696:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:706:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:717:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:724:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:731:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:738:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:745:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:752:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:759:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:766:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:773:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:780:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:787:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:794:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:801:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:808:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:815:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:822:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:829:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:836:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC ivdep
            ^
src/eigenwave3d.cpp:92:13: warning: unused variable 'rho' [-Wunused-variable]
const float rho = 1.0;
            ^
src/eigenwave3d.cpp:87:11: warning: unused variable 'margin' [-Wunused-variable]
const int margin = 2;
          ^
src/eigenwave3d.cpp:86:13: warning: unused variable 'dt' [-Wunused-variable]
const float dt = 0.002;
            ^
src/eigenwave3d.cpp:89:11: warning: unused variable 'vec_size' [-Wunused-variable]
const int vec_size = 2*dim1*dim2*dim3;
          ^
src/eigenwave3d.cpp:935:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 2);
      ^
src/eigenwave3d.cpp:933:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 1.5);
      ^
src/eigenwave3d.cpp:931:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 1.5);
      ^
src/eigenwave3d.cpp:947:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 1.5);
      ^
src/eigenwave3d.cpp:945:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 1.5);
      ^
src/eigenwave3d.cpp:943:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 2);
      ^
src/eigenwave3d.cpp:959:7: warning: unused variable 'z' [-Wunused-variable]
float z= dx3*(_z - 1.5);
      ^
src/eigenwave3d.cpp:957:7: warning: unused variable 'y' [-Wunused-variable]
float y= dx2*(_y - 2);
      ^
src/eigenwave3d.cpp:955:7: warning: unused variable 'x' [-Wunused-variable]
float x= dx1*(_x - 1.5);
      ^
src/eigenwave3d.cpp:868:5: warning: unused variable 't0' [-Wunused-variable]
int t0 = 0;
    ^
src/eigenwave3d.cpp:877:13: warning: unused variable 'rho' [-Wunused-variable]
const float rho = 1.0;
            ^
src/eigenwave3d.cpp:869:5: warning: unused variable 't1' [-Wunused-variable]
int t1 = 0;
    ^
src/eigenwave3d.cpp:870:11: warning: unused variable 'tp' [-Wunused-variable]
const int tp = 2;
          ^
src/eigenwave3d.cpp:878:13: warning: unused variable 'lambda' [-Wunused-variable]
const float lambda = 0.5;
            ^
src/eigenwave3d.cpp:872:11: warning: unused variable 'margin' [-Wunused-variable]
const int margin = 2;
          ^
src/eigenwave3d.cpp:871:13: warning: unused variable 'dt' [-Wunused-variable]
const float dt = 0.002;
            ^
src/eigenwave3d.cpp:873:11: warning: unused variable 'ntsteps' [-Wunused-variable]
const int ntsteps = 500;
          ^
src/eigenwave3d.cpp:874:11: warning: unused variable 'vec_size' [-Wunused-variable]
const int vec_size = 2*dim1*dim2*dim3;
          ^
114 warnings generated.
ld: library not found for -lgomp
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Compiling: g++ -Wall -std=c++11 -I/Users/ggorman/projects/opesci-fd/opesci/include -g -O3 -fno-tree-vectorize -fopenmp -o src/eigenwave3d src/eigenwave3d.cpp -lopesci -Wl,-rpath,/Users/ggorman/projects/opesci-fd/opesci/lib -L/Users/ggorman/projects/opesci-fd/opesci/lib

**Comment by NathanYukai** [comment]
It may took some time , but I'll try.


Date: Tue, 8 Sep 2015 06:55:54 -0700

**Comment by NathanYukai** [comment]
Added testing for  clang compiler on travis ci

**Comment by mlange05** [comment]
As discussed in the meeting, Travis is green, so this is goo to go. If anyone experiences troubles on a Mac, please raise a separate issue.

**Pull request merged**